### ISSUE TITLE
arduino: Race condition in sys/arduino/Makefile.include

### ISSUE THREAD

**Issue by gebart** [RIOT-OS/RIOT#5848]
https://ci.riot-labs.de/RIOT-OS/RIOT/5794/291c1a54ebfde8e5d7ff3db97d5b01befae3b7f0/output.html#cortex_m3_1
observed in #5794 

https://github.com/RIOT-OS/RIOT/blob/master/sys/arduino/Makefile.include#L6 is executed in parallel for all targets of a make buildtest

https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/arduino/pre_build.sh#L17 does not write the file atomically, which means that the file may be truncated at a time when the compiler from one of the boards tries to read it.

Suggested solution: Create the _sketches.cpp file inside `${BINDIR}` and create it as a part of the build process, not during evaluation of the Makefile. Also, write _sketches.cpp atomically by building a temporary file first, then renaming to the real file.

**Comment by kYc0o** [comment]
@gebart Did you have time to work on the solution? @kaspar030 any other suggestion?

**Comment by gebart** [comment]
I don't think I will have any time to spend on this issue, it is of very low priority for my use cases.

**Comment by cladmi** [comment]
To fix that correctly by generating something in `BINDIR`, I think the best would be to create an `$(APPLICATION)_sketches` module in BINDIR.

I have use cases where the RIOT repository should be read-only and build are be produced in directories outside, so it definitely makes sense to keep APPDIR as is.

My reasons for thinking about a module is that, from what I understood, `Makefile.base` is meant to build from sources files that are in the current directory.
https://github.com/RIOT-OS/RIOT/blob/2017.07-branch/Makefile.base#L86
So to move them out, would require adding there location to `DIR` to build them from there.

So inside `Makefile.include` create a module in $(BINDIR) with the sketches renames as .cpp and add this as USEMODULE.

I will try something to see if it works.

**Comment by cladmi** [comment]
I updated the pull request in https://github.com/RIOT-OS/RIOT/pull/7654

**Comment by tcschmidt** [comment]
@cladmi what about this one: move or close?

**Comment by cladmi** [comment]
As said in the previous comment, I updated the pull request.
